tests/core_widgets/test_presets_widget.py

```diff
@@ -0,0 +1,33 @@
+from pymmcore_plus import CMMCorePlus
+
+from micromanager_gui._core_widgets._presets_widget import PresetsWidget
+
+
+def test_preset_widget(qtbot, global_mmcore: CMMCorePlus):
+
+    groups = list(global_mmcore.getAvailableConfigGroups())
+
+    for group in groups:
+
+        presets = list(global_mmcore.getAvailableConfigs(group))
+
+        if len(presets) <= 1:
+            return
+
```

Comment:
```suggestion
```
any reason not to test this?  If `group` is in fact a valid configuration group name, then the user shouldn't have to always check whether that group has greater than preset before creating `PresetsWidget(group)`.  It should just show an empty (or 1-item) combobox.

Reply:
Along those lines, we could consider handling the systemConfigurationLoaded event here and refresh the combo box choices (or, if the group is no longer present, disable the combobox and add a stub item that says "No Configuration group named...")

The goal for all of these "generic" widgets is to be as accurate of a view onto the state of mmcore as possible.  We can decide how we'd like to handle incompatible changes to core (such as the configuration group disappearing)... but one way or the other these widgets should either reflect the state of core at all times, or fail loudly.

Occasionally, we'll run into situations where mmcore itself simply doesn't emit the signal(s) we need to stay in sync, in which case we can try to fix upstream.  But short of that we should evaluate all the possible model changes that might necessitate a view change.  (This can also be an iterative process, and doesn't need to happen all in one PR)

Reply:
> any reason not to test this?

I put this here just for the tests...since I wanted to test both the mmcore.events.configSet signal 
```
with qtbot.waitSignal(global_mmcore.events.configSet):
            global_mmcore.setConfig(group, presets[1])
        assert wdg.value() == presets[1]
```
and the signal emitted by the widget

```
wdg.setValue(presets[0])
assert global_mmcore.getCurrentConfig(group) == presets[0]
```
And I thought that if there is only 1 preset (e.g. in the `System` group configuration) I could not use `preset[1]`


Do you think this is better:

```
def test_preset_widget(qtbot, global_mmcore: CMMCorePlus):

    groups = list(global_mmcore.getAvailableConfigGroups())

    for group in groups:

        presets = list(global_mmcore.getAvailableConfigs(group))

        wdg = PresetsWidget(group)
        qtbot.addWidget(wdg)
        wdg.show()

        assert list(wdg.allowedValues()) == presets

        p = presets[0] if len(presets) == 1 else presets[1]

        with qtbot.waitSignal(global_mmcore.events.configSet):
            global_mmcore.setConfig(group, p)
        assert wdg.value() == p

        wdg.setValue(presets[0])
        assert global_mmcore.getCurrentConfig(group) == presets[0]

        wdg._disconnect()
        # once disconnected, core changes shouldn't call out to the widget
        if len(presets) > 1:
            global_mmcore.setConfig(group, p)
            assert global_mmcore.getCurrentConfig(group) != wdg.value()
```

Reply:
> And I thought that if there is only 1 preset (e.g. in the System group configuration) I could not use preset[1]

true, but you can handle that farther down.  In other words, first create a widget (and make sure the widget itself can be created.  If you determine that the widget won't have any more than 1 option, then yes, it doesn't make sense to test changing the option.

Reply:
>         p = presets[0] if len(presets) == 1 else presets[1]

this still will give an index error in your test if there are no presets.  Do something like this:

```python
        wdg = PresetsWidget(group)
        qtbot.addWidget(wdg)
        presets = global_mmcore.getAvailableConfigs(group)
        assert list(wdg.allowedValues()) == presets
        if len(presets) < 1:
            return # no need testing the changes of a config group that has <1 item

        # change to the last preset
        with qtbot.waitSignal(global_mmcore.events.configSet):
            global_mmcore.setConfig(group, presets[-1])
        assert wdg.value() == presets[-1] == global_mmcore.getCurrentConfig()

```

Reply:
@tlambert03 

> Along those lines, we could consider handling the systemConfigurationLoaded event here and refresh the combo box choices (or, if the group is no longer present, disable the combobox and add a stub item that says "No Configuration group named...")

I definitely agree! I tested something like this and it work well when a new cfg is loaded.

```
def refresh(self) -> None:
        with signals_blocked(self._combo):
            self._combo.clear()
            if self._group not in self._mmc.getAvailableConfigGroups():
                self._combo.addItem(f'No group named {self._group}.')
                self._combo.setEnabled(False)
            else:
                presets = self._mmc.getAvailableConfigs(self._group)
                self._combo.addItems(presets)
                self._combo.setEnabled(True)
```

However I think it would be good to also have signals emitted when a `group` is `deleted` or when a `preset` is `deleted`. What do you think? Do you think we can add them in `pymmcore-plus`?

Reply:
> However I think it would be good to also have signals emitted when a group is deleted or when a preset is deleted. What do you think? Do you think we can add them in pymmcore-plus?

I think this would be a really set of signals to open and issue on `mmCoreAndDevices` about. You want to have a go at that? Feel free to tag me at least so I can follow along